ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -613,4 +630,32 @@ private[kvutils] object TransactionCommitter {
   def getContractState(commitContext: CommitContext, key: DamlStateKey): DamlContractState =
     commitContext.get(key).getOrElse(throw Err.MissingInputState(key)).getContractState
 
+  @SuppressWarnings(Array("org.wartremover.warts.Option2Iterable"))
+  private def transactionMinRecordTime(
+      submissionTime: Instant,
+      ledgerTime: Instant,
+      maybeDeduplicateUntil: Option[Instant],
+      timeModel: TimeModel): Instant =
+    List(
+      maybeDeduplicateUntil
+        .map(_.plus(TimeModel.Resolution)), // DeduplicateUntil defines a rejection window, endpoints inclusive
+      Some(timeModel.minRecordTime(ledgerTime)),
+      Some(timeModel.minRecordTime(submissionTime))
```

Comment:
As explained above, for pre-execution we need to ensure that the submission is still valid at the record time determined at post-execution hence the use of submission time here (we verify submission time against record time normally as part of the validation steps performed by `TransactionCommitter` in case of normal execution).

Reply:
Ok, I see - the validity of submission time is somewhat loosly defined by https://github.com/digital-asset/daml/blob/master/daml-lf/spec/contract-id.rst#submission-time, and you chose to use the ledger time model bounds for submission time check.

Reply:
Exactly!